ospf6d: Fix External routes ECMP

chiragshah6 opened the pull request:
Handle RFC 2328 16.4 Calculating AS external routes with ECMP

For ASBR (External type-1 type-2) route, if it is learn via new LSA and contains
different next-hop list. First lookup route in ospf6 route table
if it exists, merge next-hop list to existing and call the callback
to install into FIB (zebra). Delete created new route as it is
identical to existing entry in route table. 

Store sorted next-hops in route list. 

During route delete ensure if there are multiple equal cost paths only delete path and its next-hop list and update zebra. 

Ticket:CM-16139
Testing Done:
Run two ASBR with 2 ECMP paths from each
DUT neighbor receievs 4 ECMP path to a external route.
ospf6 installs all 4 ECMP path to FIB/RIB

Signed-off-by: Chirag Shah <chirag@cumulusnetworks.com>

qlyoung commented:
> WARNING: Missing a blank line after declarations
> #121: FILE: f2/ospf6_area.c:121:
> +	struct ospf6_route *copy = ospf6_route_copy(route);
> +	ospf6_route_add(copy, ospf6->route_table);
< WARNING: struct  should normally be const
< #184: FILE: f1/ospf6_asbr.c:184:
---
> #183: FILE: f2/ospf6_asbr.c:183:
--
< WARNING: Block comments use a trailing */ on a separate line
< #265: FILE: f1/ospf6_asbr.c:265:
< +	 * list along with addition of ECMP path. */
< 
< WARNING: struct  should normally be const
< #269: FILE: f1/ospf6_asbr.c:269:
< +		struct listnode *anode;
< 
< WARNING: struct  should normally be const
< #270: FILE: f1/ospf6_asbr.c:270:
< +		struct listnode *nnode, *rnode, *rnext;
< 
< WARNING: struct  should normally be const
< #271: FILE: f1/ospf6_asbr.c:271:
< +		struct ospf6_nexthop *nh, *rnh;
< 
--
< WARNING: Block comments use a trailing */ on a separate line
< #275: FILE: f1/ospf6_asbr.c:275:
< +		 * if it exist simply replace nh_list from asbr_entry */
< 
< WARNING: Block comments use a trailing */ on a separate line
< #286: FILE: f1/ospf6_asbr.c:286:
< +		 * Otherwise replace existing path's nh_list */
< 
< WARNING: suspect code indent for conditional statements (40, 56)
< #314: FILE: f1/ospf6_asbr.c:314:
< +					if (!ospf6_nexthop_is_same(rnh, nh))
< +							continue;
--
< WARNING: struct  should normally be const
< #350: FILE: f1/ospf6_asbr.c:350:
---
> #256: FILE: f2/ospf6_asbr.c:256:
--
< WARNING: Block comments use a trailing */ on a separate line
< #390: FILE: f1/ospf6_asbr.c:390:
< +		 * matching path and update effective route's  nh list. */
< 
< WARNING: struct  should normally be const
< #392: FILE: f1/ospf6_asbr.c:392:
< +			struct listnode *anode, *anext;
< 
< WARNING: struct  should normally be const
< #393: FILE: f1/ospf6_asbr.c:393:
< +			struct listnode *nnode, *rnode, *rnext;
< 
< WARNING: struct  should normally be const
< #394: FILE: f1/ospf6_asbr.c:394:
< +			struct ospf6_nexthop *nh, *rnh;
< 
< WARNING: struct  should normally be const
< #395: FILE: f1/ospf6_asbr.c:395:
< +			struct ospf6_path *o_path;
< 
< WARNING: Block comments use a trailing */ on a separate line
< #400: FILE: f1/ospf6_asbr.c:400:
< +			 * replace from paths list. */
< 
< WARNING: Block comments use a trailing */ on a separate line
< #420: FILE: f1/ospf6_asbr.c:420:
< +				 * the route's nh_list */
< 
< WARNING: Too many leading tabs - consider code refactoring
< #425: FILE: f1/ospf6_asbr.c:425:
< +						if (!ospf6_nexthop_is_same(rnh,
< 
< WARNING: Block comments use a trailing */ on a separate line
< #446: FILE: f1/ospf6_asbr.c:446:
< +				 * paths list head. */
< 
< WARNING: struct  should normally be const
< #450: FILE: f1/ospf6_asbr.c:450:
< +					struct ospf6_path *h_path;
< 
< WARNING: struct  should normally be const
< #452: FILE: f1/ospf6_asbr.c:452:
< +					h_path = (struct ospf6_path *)
< 
< WARNING: struct  should normally be const
< #364: FILE: f1/ospf6_route.c:364:
< +void ospf6_path_free(struct ospf6_path *op)
< 
< WARNING: struct  should normally be const
< #371: FILE: f1/ospf6_route.c:371:
< +struct ospf6_path *ospf6_path_dup(struct ospf6_path *path)
< 
< WARNING: struct  should normally be const
< #373: FILE: f1/ospf6_route.c:373:
< +	struct ospf6_path *new;
< 
< WARNING: struct  should normally be const
< #375: FILE: f1/ospf6_route.c:375:
< +	new = XCALLOC(MTYPE_OSPF6_PATH, sizeof(struct ospf6_path));
< 
< WARNING: struct  should normally be const
< #376: FILE: f1/ospf6_route.c:376:
< +	memcpy(new, path, sizeof(struct ospf6_path));
< 
< WARNING: struct  should normally be const
< #384: FILE: f1/ospf6_route.c:384:
---
> #354: FILE: f2/ospf6_route.c:354:
--
< WARNING: struct  should normally be const
< #514: FILE: f1/ospf6_route.c:514:
---
> #473: FILE: f2/ospf6_route.c:473:
< WARNING: struct  should normally be const
< #101: FILE: f1/ospf6_route.h:101:
< +	struct list *nh_list;
< 
--
< WARNING: struct  should normally be const
< #159: FILE: f1/ospf6_route.h:159:
---
> #153: FILE: f2/ospf6_route.h:153:
--
< WARNING: struct  should normally be const
< #266: FILE: f1/ospf6_route.h:266:
---
> #259: FILE: f2/ospf6_route.h:259:
--
< WARNING: struct  should normally be const
< #341: FILE: f1/ospf6_route.h:341:
< +extern void ospf6_path_free(struct ospf6_path *op);
< 
< WARNING: struct  should normally be const
< #342: FILE: f1/ospf6_route.h:342:
< +extern struct ospf6_path *ospf6_path_dup(struct ospf6_path *path);
<

chiragshah6 commented:
@mwinter-osr  I have tried ospf6 topotest and found that 2001:db8:3::/64 has two ECMP paths (with cost of 20) with same nexthop (r4-eth0) at r4 router.  I have disabled redistribute on r1 which was the initial adv-router, at r4 the adv-router changes to r2. 
With this PR change, it is expected to combine same cost route to merge all nexthops from all available path. 
Thus, r4 you would see 3 entries rather than 4 for 2001:db8:3::/64. 


.mininet> r4 vtysh -c "show ipv6 ospf6 route 2001:db8:03::/64"
Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:01:08 ago
  Changed Time: 00:01:08 ago
Lock: 2 Flags: BA--
Memory: prev: 0x56483a0c4300 this: 0x56483a0c4620 next: 0x56483a0c4db0
Associated Area: 0.0.0.1
Path Type: Inter-Area
LS Origin: Inter-Prefix Id: 0.0.0.1 Adv: 10.0.255.3
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 0
Metric: 20 (0)
Nexthop count: 1
Nexthop:
  fe80::4806:feff:feff:1b5c r4-eth0

Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:01:08 ago
  Changed Time: 00:01:08 ago
Lock: 2 Flags: -A--
Memory: prev: 0x56483a0c4620 this: 0x56483a0c4db0 next: 0x56483a0c3e20
Associated Area: 0.0.0.1
Path Type: External-1
LS Origin: AS-External Id: 0.0.0.0 Adv: 10.0.255.3
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 1
Metric: 10 (0)
Nexthop count: 1
Nexthop:
  fe80::4806:feff:feff:1b5c r4-eth0

Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:01:08 ago
  Changed Time: 00:01:08 ago
Lock: 2 Flags: -A--
Memory: prev: 0x56483a0c4db0 this: 0x56483a0c3e20 next: 0x56483a0c48b0
Associated Area: 0.0.0.1
Path Type: External-1
LS Origin: AS-External Id: 0.0.0.0 **Adv: 10.0.255.1**
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 1
**Metric: 20 (0)**
Nexthop count: 1
Nexthop:
  fe80::4806:feff:feff:1b5c r4-eth0

mininet> r1 vtysh

Hello, this is FRRouting (version 3.1-dev).
Copyright 1996-2005 Kunihiro Ishiguro, et al.

chirag-VirtualBox# config t
chirag-VirtualBox(config)# router ospf6
chirag-VirtualBox(config-ospf6)# **no redistribute connected**
chirag-VirtualBox(config-ospf6)# end
chirag-VirtualBox# exit
mininet> r4 vtysh -c "show ipv6 ospf6 route 2001:db8:03::/64"
Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:04:24 ago
  Changed Time: 00:04:24 ago
Lock: 2 Flags: BA--
Memory: prev: 0x56483a0c4300 this: 0x56483a0c4620 next: 0x56483a0c4db0
Associated Area: 0.0.0.1
Path Type: Inter-Area
LS Origin: Inter-Prefix Id: 0.0.0.1 Adv: 10.0.255.3
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 0
Metric: 20 (0)
Nexthop count: 1
Nexthop:
  fe80::4806:feff:feff:1b5c r4-eth0

Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:04:24 ago
  Changed Time: 00:04:24 ago
Lock: 2 Flags: -A--
Memory: prev: 0x56483a0c4620 this: 0x56483a0c4db0 next: 0x56483a0c3e20
Associated Area: 0.0.0.1
Path Type: External-1
LS Origin: AS-External Id: 0.0.0.0 Adv: 10.0.255.3
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 1
Metric: 10 (0)
Nexthop count: 1
Nexthop:
  fe80::4806:feff:feff:1b5c r4-eth0

Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:04:24 ago
  Changed Time: 00:04:24 ago
Lock: 2 Flags: -A--
Memory: prev: 0x56483a0c4db0 this: 0x56483a0c3e20 next: 0x56483a0c48b0
Associated Area: 0.0.0.1
Path Type: External-1
LS Origin: AS-External Id: 0.0.0.1 **Adv: 10.0.255.2**
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 1
**Metric: 20 (0)**
Nexthop count: 0
Nexthop:

mininet>

mwinter-osr commented:
Something doesn't look right with the ecmp changed:

R4 looks like this:
```r4 vtysh -c "show ipv6 ospf6 route"
*N IE 2001:db8:1::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
 N E1 2001:db8:1::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
*N IE 2001:db8:2::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
 N E1 2001:db8:2::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
*N IE 2001:db8:3::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
 N E1 2001:db8:3::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
 N E1 2001:db8:3::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
*N IE 2001:db8:100::/64              fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
 N E1 2001:db8:100::/64              fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
*N IA 2001:db8:200::/64              ::                        r4-eth0 00:14:29
 N E1 2001:db8:200::/64              fe80::e45e:83ff:fecc:1904 r4-eth0 00:14:24
*N IA 2001:db8:300::/64              ::                        r4-eth1 00:14:29
```

And the detailed view shows the `E1 2001:db8:3::/64` route with the advertising host r2 (10.0.255.2) was removed (compressed) as r1 and r2 had the same metric (20).

Now, if I go to r1 (the one displayed) and shutdown the interface (`ifconfig r1-eth1 down`), I would have expected to see the same number of routes (there is now no route to optimize away).

However, the routes look like this:
```r4 vtysh -c "show ipv6 ospf6 route"
*N IE 2001:db8:2::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:18:50
 N E1 2001:db8:2::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:18:50
*N IE 2001:db8:3::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:18:50
 N E1 2001:db8:3::/64                fe80::e45e:83ff:fecc:1904 r4-eth0 00:18:50
*N IE 2001:db8:100::/64              fe80::e45e:83ff:fecc:1904 r4-eth0 00:18:50
 N E1 2001:db8:100::/64              fe80::e45e:83ff:fecc:1904 r4-eth0 00:18:50
*N IA 2001:db8:200::/64              ::                        r4-eth0 00:18:55
 N E1 2001:db8:200::/64              fe80::e45e:83ff:fecc:1904 r4-eth0 00:18:50
*N IA 2001:db8:300::/64              ::                        r4-eth1 00:18:55
```

And looking at the details, I see (for 2001:db8:3::/64):
```Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:19:04 ago
  Changed Time: 00:19:04 ago
Lock: 2 Flags: BA--
Memory: prev: 0x55a6dffc4270 this: 0x55a6dffc48d0 next: 0x55a6dffc4d20
Associated Area: 0.0.0.1
Path Type: Inter-Area
LS Origin: Inter-Prefix Id: 0.0.0.2 Adv: 10.0.255.3
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 0
Metric: 20 (0)
Nexthop count: 1
Nexthop:
  fe80::e45e:83ff:fecc:1904 r4-eth0

Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:19:04 ago
  Changed Time: 00:19:04 ago
Lock: 2 Flags: -A--
Memory: prev: 0x55a6dffc48d0 this: 0x55a6dffc4d20 next: 0x55a6dffc3fc0
Associated Area: 0.0.0.1
Path Type: External-1
LS Origin: AS-External Id: 0.0.0.0 Adv: 10.0.255.3
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 1
Metric: 10 (0)
Nexthop count: 1
Nexthop:
  fe80::e45e:83ff:fecc:1904 r4-eth0

Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:19:04 ago
  Changed Time: 00:19:04 ago
Lock: 2 Flags: -A--
Memory: prev: 0x55a6dffc4d20 this: 0x55a6dffc3fc0 next: 0x55a6dffc4620
Associated Area: 0.0.0.1
Path Type: External-1
LS Origin: AS-External Id: 0.0.0.0 Adv: 10.0.255.2
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 1
Metric: 20 (0)
Nexthop count: 0
Nexthop:
```

(Yes, not a mistake! The last route has no nexthop!)

This looks wrong to me.

chiragshah6 commented:
@mwinter-osr 
I have addressed the issue in latest patch, where two of the ECMP paths having same nexthop. When one of the path removed, it deleted associated nexthop list from the route. 
 Also added print in detail output showing number of paths

r4 state: 
Before r1 r1-eth1 down 
{code}
Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:00:34 ago
  Changed Time: 00:00:34 ago
Lock: 2 Flags: -A--
Memory: prev: 0x55eb0e0ade60 this: 0x55eb0e0accb0 next: 0x55eb0e0ad4d0
Associated Area: 0.0.0.1
Path Type: External-1
LS Origin: AS-External Id: 0.0.0.1 Adv: **10.0.255.1**
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx
Metric Type: 1
Metric: 20 (0)
**Paths count: 2**
Nexthop count: 1
Nexthop:
  **fe80::2c9b:b9ff:febd:d284 r4-eth0**
{code}

After ifconfig r1-eth1 down at r1, below is r4 state. 
{code}
Destination: 2001:db8:3::/64
Destination type: Network
Installed Time: 00:11:11 ago 
  Changed Time: 00:11:11 ago 
Lock: 2 Flags: -A--
Memory: prev: 0x55eb0e0ade60 this: 0x55eb0e0accb0 next: 0x55eb0e0ad4d0
Associated Area: 0.0.0.1
Path Type: External-1
LS Origin: AS-External Id: 0.0.0.0 Adv: **10.0.255.2**
Options: --|-|-|--|-|--
Router Bits: --------
Prefix Options: xxx 
Metric Type: 1
Metric: 20 (0) 
**Paths count: 1**
Nexthop count: 1
Nexthop:
  **fe80::2c9b:b9ff:febd:d284 r4-eth0**
{code}

mwinter-osr commented:
Trying to further test, but tests fail because of unrelated crash in zebra (Issue #1575)

Issue #1575 is a teststopper for this. I'll continue testing after the issue is resolved.

Jafaral commented:
@mwinter-osr  is this test issue resolved?

mwinter-osr commented:
Manual verification looks good.

I'll like to update topotests to not flag it as a fail, so please give me a little more time to modify topotests to allow old and new format to pass.

Please wait with merging - I'll update as soon as I have topotests adapted (may take another 24hrs)

mwinter-osr commented:
All good from my side & Topotests updated, so it all passes.
I see a outstanding review from @Jafaral , not sure if he or someone else wants to review it.

Jafaral commented:
Thanks @mwinter-osr  I will give it a go
